Close #128: Multiple networks in the same binary

LeeSmet opened the pull request:
Some of the types variables are still initialized instead of only declared since the tests don't seem to like it otherwise

GlenDC commented:
Will review later, however I can already tell you that you will have to rebase our branch and resolve some conflicts, as the rivinec/rivined packages have been renamed and moved.

LeeSmet commented:
weird since I pulled master and created my branch just before I committed these changed and pushed them

GlenDC commented:
> weird since I pulled master and created my branch just before I committed these changed and pushed them

It's because I pushed those changes when you just made the PR, I saw that too late. Either way, pushed those changes as early as possible, as I didn't want them to impact all those config-related features we'll implement over the next couple of days.

GlenDC commented on types/constants.go:
```diff
@@ -50,200 +50,153 @@ var (
+// are set, since some (e.g. GenesisTimestamp) are chain specific, and this also
+// allows some santiy checking later
+func DefaultChainConstants() ChainConstants {
+	// GenesisTimestamp, GenesisBlockStakeAllocation, and GenesisCoinDistribution aren't set as there is no such thing as a "sane default" for these variables
```
Should just be a godoc as part of the comment above the function. So it's very visible, as it concerns the user for sure.

GlenDC commented on types/constants.go:
```diff
@@ -50,200 +50,153 @@ var (
+type ChainConstants struct {
+	BlockSizeLimit   uint64
+	RootDepth        Target
+	BlockFrequency   BlockHeight
```
Please provide properties for each property, explaining what they do

GlenDC commented on pkg/daemon/networks.go:
```diff
@@ -0,0 +1,44 @@
+)
+
+// Network are variables for a particular chain. Currently, these are genesis constants and bootstrap peers
+type Network struct {
```
I would call that NetworkConfig.

LeeSmet commented:
Added a WIP for feedback: 

- Replaced the exported global vars in the types package with an unexported cfg struct as some of the function in types rely on it.
- Inject the chain constants struct in all modules that require them (basically all)
- Inject the bootstrappeers in the gateway module
- Fix tests (fml)

For now I've replaced the onecoin reference in client with an injected var when the client is created since the chainconstants are now defined in the context of a network, which is defined only in the daemon. Will check tomorrow if there is a better solution for this but at least it compiles for now

GlenDC commented:
I'll make sure this is today around lunch-afternoon merged, as I think this is the last remaining thing of milestone `0.5`.
